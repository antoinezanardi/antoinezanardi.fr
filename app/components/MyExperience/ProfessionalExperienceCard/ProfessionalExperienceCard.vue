<template>
  <div class="card">
    <div class="row">
      <div
        class="bg-primary col-md-3"
        data-aos="fade-right"
        data-aos-duration="500"
        data-aos-offset="50"
      >
        <div class="card-body cc-experience-header">
          <PeriodTimeline
            class="experience-timeline"
            :finished-at="professionalExperience.job.finishedAt"
            :image="professionalExperience.company.image"
            :image-alt="professionalExperience.company.name"
            :started-at="professionalExperience.job.startedAt"
            :url="professionalExperience.company.url"
          />
        </div>
      </div>

      <div
        class="col-md-9"
        data-aos="fade-left"
        data-aos-duration="500"
        data-aos-offset="50"
      >
        <div class="card-body">
          <h4
            class="font-weight-bold h4 job-name mt-0"
          >
            {{ professionalExperience.job.name }}
          </h4>

          <hr>

          <p
            v-for="(paragraph, index) in professionalExperience.job.description"
            :key="index"
            class="job-description"
          >
            {{ paragraph }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import type { ProfessionalExperienceCardProps } from "~/components/MyExperience/ProfessionalExperienceCard/professional-experience-card.types";
import PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";

defineProps<ProfessionalExperienceCardProps>();
</script>

<style lang="scss" scoped>
.cc-experience-header {
  padding-top: 25px !important;
}
@media (max-width: 768px) {
  .cc-experience-header {
    padding-right: 1.25rem !important;
  }
}
</style>