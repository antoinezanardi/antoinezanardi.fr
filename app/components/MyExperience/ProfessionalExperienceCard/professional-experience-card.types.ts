import type { ProfessionalExperience } from "~/models/ProfessionalExperience";

type ProfessionalExperienceCardProps = {
  professionalExperience: ProfessionalExperience;
};

export type { ProfessionalExperienceCardProps };