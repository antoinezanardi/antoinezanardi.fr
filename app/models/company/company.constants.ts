import type { CompanyName, Company } from "~/models/company/company.types";

const COMPANY_NAMES = [
  "SoBook",
  "OhMyCode",
  "Daveo",
] as const;

const COMPANIES: Record<CompanyName, Company> = Object.freeze({
  OhMyCode: {
    name: "OhMyCode",
    url: "https://www.sobook.fr/oh-my-code/",
    image: "omc-logo.png",
  },
  SoBook: {
    name: "SoBook",
    url: "https://www.sobook.fr/oh-my-code/",
    image: "sobook-logo.png",
  },
  Daveo: {
    name: "Daveo",
    url: "https://www.daveo.fr/",
    image: "daveo-logo.png",
  },
});

export {
  COMPANY_NAMES,
  COMPANIES,
};