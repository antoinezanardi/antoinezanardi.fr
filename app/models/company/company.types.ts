import type { TupleToUnion } from "type-fest";

import type { COMPANY_NAMES } from "~/models/company/company.constants";

type CompanyName = TupleToUnion<typeof COMPANY_NAMES>;

type Company = {
  name: CompanyName;
  image: string;
  url: string;
};

export type { Company, CompanyName };