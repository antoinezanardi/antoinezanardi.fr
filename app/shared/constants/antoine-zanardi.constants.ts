const ANTOINE_ZANARDI_FULL_NAME = "Antoine ZANARDI";

const ANTOINE_ZANARDI_BIRTH_DATE = "1996-04-14";

const ANTOINE_ZANARDI_GITHUB_URL = "https://github.com/antoinezanardi/";

const ANTOINE_ZANARDI_GITHUB_AVATAR_URL = "https://avatars.githubusercontent.com/u/11405981?v=4";

const ANTOINE_ZANARDI_LINKEDIN_URL = "https://www.linkedin.com/in/antoinezanardi/";

export {
  ANTOINE_ZANARDI_FULL_NAME,
  ANTOINE_ZANARDI_BIRTH_DATE,
  ANTOINE_ZANARDI_GITHUB_URL,
  ANTOINE_ZANARDI_GITHUB_AVATAR_URL,
  ANTOINE_ZANARDI_LINKEDIN_URL,
};