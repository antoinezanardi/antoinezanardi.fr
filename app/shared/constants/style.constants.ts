const PRIMARY_COLOR = "#12671D";

const BLACK_COLOR = "#000000";

const WHITE_COLOR = "#FFFFFF";

export {
  PRIMARY_COLOR,
  BLACK_COLOR,
  WHITE_COLOR,
};