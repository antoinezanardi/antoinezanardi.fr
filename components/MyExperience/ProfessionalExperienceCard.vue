<template>
  <div class="card">
    <div class="row">
      <div class="col-md-3 bg-primary" data-aos="fade-right" data-aos-offset="50" data-aos-duration="500">
        <div class="card-body cc-experience-header">
          <a :href="professionalExperience.company.url" target="_blank" rel="noopener noreferrer">
            <img
              class="white-logo company-logo"
              :alt="professionalExperience.company.image"
              :src="`/assets/images/${professionalExperience.company.image}`"
            />
          </a>
          <hr class="my-3"/>
          <div class="experience-date" v-html="formattedJobFinishedAt"/>
          <i class="fa fa-arrow-up fa-2x my-3"/>
          <div class="experience-date" v-html="formattedJobStartedAt"/>
          <div class="font-weight-bold font-italic small mt-4" v-html="formattedJobPeriod"/>
        </div>
      </div>
      <div class="col-md-9" data-aos="fade-left" data-aos-offset="50" data-aos-duration="500">
        <div class="card-body">
          <div class="h5" v-html="professionalExperience.job.name"/>
          <p v-for="(paragraph, index) of professionalExperience.job.description" :key="index" v-html="paragraph"/>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { useI18n } from "vue-i18n";
import { computed } from "#imports";
import { useDates } from "~/composables/useDates";
import type { ProfessionalExperience } from "~/models/ProfessionalExperience";

const props = defineProps<{
  professionalExperience: ProfessionalExperience;
}>();

const { t } = useI18n();
const { getMonthFullName, getPeriod } = useDates();

const formattedJobPeriod = computed<string>(() => {
  const finishedAt = props.professionalExperience.job.finishedAt ?? new Date();
  const period = getPeriod(props.professionalExperience.job.startedAt, finishedAt);
  let periodText = "";
  if (period.year) {
    periodText += t("shared.years", { yearCount: period.year }, period.year);
  }
  if (period.month && period.year) {
    periodText += ` ${t("shared.and")} `;
  }
  if (period.month) {
    periodText += t("shared.months", { monthCount: period.month });
  }
  return `( ${periodText} )`;
});
const formattedJobStartedAt = computed<string>(() => {
  const { startedAt } = props.professionalExperience.job;
  const monthFullName = getMonthFullName(props.professionalExperience.job.startedAt);
  return `${monthFullName} ${startedAt.getFullYear()}`;
});
const formattedJobFinishedAt = computed<string>(() => {
  const { finishedAt } = props.professionalExperience.job;
  return finishedAt ? `${getMonthFullName(finishedAt)} ${finishedAt.getFullYear()}` : t("shared.today");
});
</script>

<style lang="scss" scoped>
.cc-experience-header {
  padding-top: 25px !important;
}
@media (max-width: 768px) {
  .cc-experience-header {
    padding-right: 1.25rem !important;
  }
}
.company-logo {
  max-width: 180px;
  max-height: 75px;
  filter: brightness(0) invert(1);
}
.experience-date {
  font-size: 1.25rem;
}
</style>