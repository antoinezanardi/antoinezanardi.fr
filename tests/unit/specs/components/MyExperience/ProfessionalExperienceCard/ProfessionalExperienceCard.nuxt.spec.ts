import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { ProfessionalExperienceCardProps } from "~/components/MyExperience/ProfessionalExperienceCard/professional-experience-card.types";
import ProfessionalExperienceCard from "~/components/MyExperience/ProfessionalExperienceCard/ProfessionalExperienceCard.vue";
import type PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";
import { COMPANIES } from "~/models/company/company.constants";

describe("Professional Experience Card Component", () => {
  const defaultProps: ProfessionalExperienceCardProps = {
    professionalExperience: {
      job: {
        name: "MyExperience.internFullStackDeveloper",
        description: ["MyExperience.distributionProject", "MyExperience.thisFinalInternship"],
        startedAt: new Date("2018-09-01T00:00:00.000Z"),
        finishedAt: new Date("2019-08-30T00:00:00.000Z"),
      },
      company: COMPANIES.OhMyCode,
    },
  };
  let wrapper: ReturnType<typeof mount<typeof ProfessionalExperienceCard>>;

  async function mountProfessionalExperienceCardComponent(options: ComponentMountingOptions<typeof ProfessionalExperienceCard> = {}):
  Promise<ReturnType<typeof mount<typeof ProfessionalExperienceCard>>> {
    return mountSuspendedComponent(ProfessionalExperienceCard, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountProfessionalExperienceCardComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Period Timeline", () => {
    it("should pass job started date when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".experience-timeline");

      expect(periodTimeline.props("startedAt")).toBe(defaultProps.professionalExperience.job.startedAt);
    });

    it("should pass job finished date when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".experience-timeline");

      expect(periodTimeline.props("finishedAt")).toBe(defaultProps.professionalExperience.job.finishedAt);
    });

    it("should pass company image when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".experience-timeline");

      expect(periodTimeline.props("image")).toBe(defaultProps.professionalExperience.company.image);
    });

    it("should pass company url when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".experience-timeline");

      expect(periodTimeline.props("url")).toBe(defaultProps.professionalExperience.company.url);
    });
  });

  describe("Job", () => {
    it("should render job name when rendered.", () => {
      const jobName = wrapper.find<HTMLDivElement>(".job-name");

      expect(jobName.text()).toBe("MyExperience.internFullStackDeveloper");
    });

    it("should render job description when rendered.", () => {
      const jobDescriptions = wrapper.findAll<HTMLParagraphElement>(".job-description");

      expect(jobDescriptions).toHaveLength(2);
      expect(jobDescriptions[0].text()).toBe("MyExperience.distributionProject");
      expect(jobDescriptions[1].text()).toBe("MyExperience.thisFinalInternship");
    });
  });
});