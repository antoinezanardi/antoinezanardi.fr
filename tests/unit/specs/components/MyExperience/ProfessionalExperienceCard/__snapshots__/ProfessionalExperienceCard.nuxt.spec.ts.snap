// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Professional Experience Card Component > should match snapshot when rendered. 1`] = `
"<div data-v-ceaf3c35="" class="card">
  <div data-v-ceaf3c35="" class="row">
    <div data-v-ceaf3c35="" class="bg-primary col-md-3" data-aos="fade-right" data-aos-duration="500" data-aos-offset="50">
      <div data-v-ceaf3c35="" class="card-body cc-experience-header">
        <period-timeline-stub data-v-ceaf3c35="" image="omc-logo.png" url="https://www.sobook.fr/oh-my-code/" doesshowyearonly="false" startedat="Sat Sep 01 2018 02:00:00 GMT+0200 (Central European Summer Time)" finishedat="Fri Aug 30 2019 02:00:00 GMT+0200 (Central European Summer Time)" class="experience-timeline"></period-timeline-stub>
      </div>
    </div>
    <div data-v-ceaf3c35="" class="col-md-9" data-aos="fade-left" data-aos-duration="500" data-aos-offset="50">
      <div data-v-ceaf3c35="" class="card-body">
        <div data-v-ceaf3c35="" class="h4 job-name">MyExperience.internFullStackDeveloper</div>
        <hr data-v-ceaf3c35="">
        <p data-v-ceaf3c35="" class="job-description">MyExperience.distributionProject</p>
        <p data-v-ceaf3c35="" class="job-description">MyExperience.thisFinalInternship</p>
      </div>
    </div>
  </div>
</div>"
`;
